src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -618,6 +618,33 @@ public static bool IsNamespaceDeclarationNameContext(this SyntaxTree syntaxTree,
             return declaration != null && (declaration.Name.Span.IntersectsWith(position) || declaration.NamespaceKeyword == token);
         }
 
+        public static bool IsPartialTypeDeclarationNameContext(this SyntaxTree syntaxTree, int position, CancellationToken cancellationToken, out TypeDeclarationSyntax declarationSyntax)
+        {
+            if (syntaxTree.IsInNonUserCode(position, cancellationToken))
+            {
+                declarationSyntax = null;
+                return false;
+            }
+
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken)
+                            .GetPreviousTokenIfTouchingWord(position);
+
+            switch (token.Kind())
+            {
+                case SyntaxKind.ClassKeyword:
+                case SyntaxKind.StructKeyword:
+                case SyntaxKind.InterfaceKeyword:
+                    declarationSyntax = token.GetAncestor<TypeDeclarationSyntax>();
+                    return declarationSyntax != null &&
+                           declarationSyntax.Keyword == token &&
+                           declarationSyntax.Modifiers.Any(SyntaxKind.PartialKeyword);
```

Comment:
it's actually much simpler in C#.  "partial" has to immediately precede class/interface/struct.  So you can just walk back one token and check that.